samples/subsys/smf/smf_calculator/src/main.c

```diff
@@ -0,0 +1,188 @@
+{
+	while (k_msgq_put(&output_msgq, output, K_NO_WAIT) != 0) {
+		k_msgq_purge(&output_msgq);
+	}
+}
+
+static void lv_btn_matrix_click_callback(lv_event_t *e)
+{
+	lv_event_code_t code = lv_event_get_code(e);
+	lv_obj_t *obj = lv_event_get_target(e);
+
+	if (code == LV_EVENT_PRESSED) {
+		uint32_t id = lv_btnmatrix_get_selected_btn(obj);
+
+		if (id > ARRAY_SIZE(buttons)) {
+			LOG_ERR("Invalid button: %d", id);
+			return;
+		}
+
+		int rc = post_calculator_event(&buttons[id].event, K_FOREVER);
+
+		if (rc != 0) {
+			LOG_ERR("could not post to msgq: %d", rc);
+		}
+	}
+}
+
+int setup_display(void)
+{
+	const struct device *display_dev;
+
+	display_dev = DEVICE_DT_GET(DT_CHOSEN(zephyr_display));
+	if (!device_is_ready(display_dev)) {
+		LOG_ERR("Device not ready, aborting setup");
+		return -ENODEV;
+	}
+
+	static const char *const btnm_map[] = {buttons[BUTTON_CANCEL_ENTRY].label,
+					       buttons[BUTTON_CANCEL].label,
+					       "\n",
+					       buttons[BUTTON_7].label,
+					       buttons[BUTTON_8].label,
+					       buttons[BUTTON_9].label,
+					       buttons[BUTTON_DIVIDE].label,
+					       "\n",
+					       buttons[BUTTON_4].label,
+					       buttons[BUTTON_5].label,
+					       buttons[BUTTON_6].label,
+					       buttons[BUTTON_MULTIPLY].label,
+					       "\n",
+					       buttons[BUTTON_1].label,
+					       buttons[BUTTON_2].label,
+					       buttons[BUTTON_3].label,
+					       buttons[BUTTON_SUBTRACT].label,
+					       "\n",
+					       buttons[BUTTON_0].label,
+					       buttons[BUTTON_DECIMAL_POINT].label,
+					       buttons[BUTTON_EQUALS].label,
+					       buttons[BUTTON_ADD].label,
+					       "\n"};
+
+	lv_obj_t *btn_matrix = lv_btnmatrix_create(lv_scr_act());
+
+	lv_obj_align(btn_matrix, LV_ALIGN_BOTTOM_MID, 0, 0);
+	lv_btnmatrix_set_map(btn_matrix, (const char **)btnm_map);
+	lv_obj_set_size(btn_matrix, lv_pct(CALC_BTN_WIDTH_PCT), lv_pct(CALC_BTN_HEIGHT_PCT));
+	lv_obj_add_event_cb(btn_matrix, lv_btn_matrix_click_callback, LV_EVENT_ALL, NULL);
+
+	result_label = lv_label_create(lv_scr_act());
+	lv_obj_set_width(result_label, lv_pct(CALC_RESULT_WIDTH_PCT));
+	lv_obj_set_style_text_align(result_label, LV_TEXT_ALIGN_RIGHT, 0);
+	lv_obj_align(result_label, LV_ALIGN_TOP_MID, 0, lv_pct(CALC_RESULT_OFFSET_PCT));
+
+	static lv_style_t style_shadow;
+
+	lv_style_init(&style_shadow);
+	lv_style_set_shadow_width(&style_shadow, 5);
+	lv_style_set_shadow_spread(&style_shadow, 2);
+	lv_style_set_shadow_color(&style_shadow, lv_palette_main(LV_PALETTE_GREY));
+	lv_obj_add_style(result_label, &style_shadow, 0);
+	update_display("0");
+
+	lv_task_handler();
+	display_blanking_off(display_dev);
+
+	return 0;
+}
+
+int main(void)
+{
+	printk("SMF Desk Caclualtor Demo\n");
+
+	int rc = setup_display();
+
+	if (rc != 0) {
+		return rc;
+	}
+
+	while (1) {
+		char output[CALCULATOR_STRING_LENGTH];
+
+		if (k_msgq_get(&output_msgq, output, K_MSEC(50)) == 0) {
+			lv_label_set_text(result_label, output);
+		}
```

Comment:
do you need this queue? can't you just call `lv_label_set_text` directly?

Reply:
lvgl is not thread safe, UI updates should be done by the thread calling the task handler. (At least for 8.4, new version has locking for this sort of stuff)

Reply:
sounds good thanks for clarifying